// @genkit-ai/google-cloud/lib/telemetry/action.js
"use strict";
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var action_exports = {};
__export(action_exports, {
  actionTelemetry: () => actionTelemetry
});
module.exports = __toCommonJS(action_exports);
var import_api = require("@opentelemetry/api");
var import_core = require("@opentelemetry/core");
var import_genkit = require("genkit");
var import_logging = require("genkit/logging");
var import_tracing = require("genkit/tracing");
var import_metrics = require("../metrics.js");
var import_utils = require("../utils.js");
class ActionTelemetry {
  constructor() {
    /**
     * Wraps the declared metrics in a Genkit-specific, internal namespace.
     */
    this._N = import_metrics.internalMetricNamespaceWrap.bind(null, "action");
    this.actionCounter = new import_metrics.MetricCounter(this._N("requests"), {
      description: "Counts calls to genkit actions.",
      valueType: import_api.ValueType.INT
    });
    this.actionLatencies = new import_metrics.MetricHistogram(this._N("latency"), {
      description: "Latencies when calling Genkit actions.",
      valueType: import_api.ValueType.DOUBLE,
      unit: "ms"
    });
  }
  tick(span, paths, logIO, projectId) {
    const attributes = span.attributes;
    const actionName = attributes["genkit:name"] || "<unknown>";
    const subtype = attributes["genkit:metadata:subtype"];
    const path = attributes["genkit:path"] || "<unknown>";
    let featureName = (0, import_utils.extractOuterFeatureNameFromPath)(path);
    if (!featureName || featureName === "<unknown>") {
      featureName = actionName;
    }
    const state = attributes["genkit:state"] || "success";
    const latencyMs = (0, import_core.hrTimeToMilliseconds)(
      (0, import_core.hrTimeDuration)(span.startTime, span.endTime)
    );
    const errorName = (0, import_utils.extractErrorName)(span.events);
    if (state === "success") {
      this.writeSuccess(actionName, featureName, path, latencyMs);
    } else if (state === "error") {
      this.writeFailure(actionName, featureName, path, latencyMs, errorName);
    } else {
      import_logging.logger.warn(`Unknown action state; ${state}`);
    }
    if (subtype === "tool" && logIO) {
      const input = attributes["genkit:input"];
      const output = attributes["genkit:output"];
      if (input) {
        this.recordIO(span, "Input", featureName, path, input, projectId);
      }
      if (output) {
        this.recordIO(span, "Output", featureName, path, output, projectId);
      }
    }
  }
  writeSuccess(actionName, featureName, path, latencyMs) {
    const dimensions = {
      name: actionName,
      featureName,
      path,
      status: "success",
      source: "ts",
      sourceVersion: import_genkit.GENKIT_VERSION
    };
    this.actionCounter.add(1, dimensions);
    this.actionLatencies.record(latencyMs, dimensions);
  }
  writeFailure(actionName, featureName, path, latencyMs, errorName) {
    const dimensions = {
      name: actionName,
      featureName,
      path,
      source: "ts",
      sourceVersion: import_genkit.GENKIT_VERSION,
      status: "failure",
      error: errorName
    };
    this.actionCounter.add(1, dimensions);
    this.actionLatencies.record(latencyMs, dimensions);
  }
  recordIO(span, tag, featureName, qualifiedPath, input, projectId) {
    const path = (0, import_tracing.toDisplayPath)(qualifiedPath);
    const sharedMetadata = __spreadProps(__spreadValues({}, (0, import_utils.createCommonLogAttributes)(span, projectId)), {
      path,
      qualifiedPath,
      featureName
    });
    import_logging.logger.logStructured(`${tag}[${path}, ${featureName}]`, __spreadProps(__spreadValues({}, sharedMetadata), {
      content: input
    }));
  }
}
const actionTelemetry = new ActionTelemetry();
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  actionTelemetry
});
