// @genkit-ai/google-cloud/lib/telemetry/action.mjs
import {
  __spreadProps,
  __spreadValues
} from "../chunk-E3LOUS7X.mjs";
import { ValueType } from "@opentelemetry/api";
import { hrTimeDuration, hrTimeToMilliseconds } from "@opentelemetry/core";
import { GENKIT_VERSION } from "genkit";
import { logger } from "genkit/logging";
import { toDisplayPath } from "genkit/tracing";
import {
  MetricCounter,
  MetricHistogram,
  internalMetricNamespaceWrap
} from "../metrics.js";
import {
  createCommonLogAttributes,
  extractErrorName,
  extractOuterFeatureNameFromPath
} from "../utils.js";
class ActionTelemetry {
  constructor() {
    /**
     * Wraps the declared metrics in a Genkit-specific, internal namespace.
     */
    this._N = internalMetricNamespaceWrap.bind(null, "action");
    this.actionCounter = new MetricCounter(this._N("requests"), {
      description: "Counts calls to genkit actions.",
      valueType: ValueType.INT
    });
    this.actionLatencies = new MetricHistogram(this._N("latency"), {
      description: "Latencies when calling Genkit actions.",
      valueType: ValueType.DOUBLE,
      unit: "ms"
    });
  }
  tick(span, paths, logIO, projectId) {
    const attributes = span.attributes;
    const actionName = attributes["genkit:name"] || "<unknown>";
    const subtype = attributes["genkit:metadata:subtype"];
    const path = attributes["genkit:path"] || "<unknown>";
    let featureName = extractOuterFeatureNameFromPath(path);
    if (!featureName || featureName === "<unknown>") {
      featureName = actionName;
    }
    const state = attributes["genkit:state"] || "success";
    const latencyMs = hrTimeToMilliseconds(
      hrTimeDuration(span.startTime, span.endTime)
    );
    const errorName = extractErrorName(span.events);
    if (state === "success") {
      this.writeSuccess(actionName, featureName, path, latencyMs);
    } else if (state === "error") {
      this.writeFailure(actionName, featureName, path, latencyMs, errorName);
    } else {
      logger.warn(`Unknown action state; ${state}`);
    }
    if (subtype === "tool" && logIO) {
      const input = attributes["genkit:input"];
      const output = attributes["genkit:output"];
      if (input) {
        this.recordIO(span, "Input", featureName, path, input, projectId);
      }
      if (output) {
        this.recordIO(span, "Output", featureName, path, output, projectId);
      }
    }
  }
  writeSuccess(actionName, featureName, path, latencyMs) {
    const dimensions = {
      name: actionName,
      featureName,
      path,
      status: "success",
      source: "ts",
      sourceVersion: GENKIT_VERSION
    };
    this.actionCounter.add(1, dimensions);
    this.actionLatencies.record(latencyMs, dimensions);
  }
  writeFailure(actionName, featureName, path, latencyMs, errorName) {
    const dimensions = {
      name: actionName,
      featureName,
      path,
      source: "ts",
      sourceVersion: GENKIT_VERSION,
      status: "failure",
      error: errorName
    };
    this.actionCounter.add(1, dimensions);
    this.actionLatencies.record(latencyMs, dimensions);
  }
  recordIO(span, tag, featureName, qualifiedPath, input, projectId) {
    const path = toDisplayPath(qualifiedPath);
    const sharedMetadata = __spreadProps(__spreadValues({}, createCommonLogAttributes(span, projectId)), {
      path,
      qualifiedPath,
      featureName
    });
    logger.logStructured(`${tag}[${path}, ${featureName}]`, __spreadProps(__spreadValues({}, sharedMetadata), {
      content: input
    }));
  }
}
const actionTelemetry = new ActionTelemetry();
export {
  actionTelemetry
};
