// @genkit-ai/google-cloud/lib/telemetry/feature.mjs
import {
  __spreadProps,
  __spreadValues
} from "../chunk-E3LOUS7X.mjs";
import { ValueType } from "@opentelemetry/api";
import { hrTimeDuration, hrTimeToMilliseconds } from "@opentelemetry/core";
import { GENKIT_VERSION, GenkitError } from "genkit";
import { logger } from "genkit/logging";
import { toDisplayPath } from "genkit/tracing";
import {
  MetricCounter,
  MetricHistogram,
  internalMetricNamespaceWrap
} from "../metrics.js";
import { createCommonLogAttributes, extractErrorName } from "../utils.js";
class FeaturesTelemetry {
  constructor() {
    /**
     * Wraps the declared metrics in a Genkit-specific, internal namespace.
     */
    this._N = internalMetricNamespaceWrap.bind(null, "feature");
    this.featureCounter = new MetricCounter(this._N("requests"), {
      description: "Counts calls to genkit features.",
      valueType: ValueType.INT
    });
    this.featureLatencies = new MetricHistogram(this._N("latency"), {
      description: "Latencies when calling Genkit features.",
      valueType: ValueType.DOUBLE,
      unit: "ms"
    });
  }
  tick(span, paths, logIO, projectId) {
    const attributes = span.attributes;
    const name = attributes["genkit:name"];
    const path = attributes["genkit:path"];
    const latencyMs = hrTimeToMilliseconds(
      hrTimeDuration(span.startTime, span.endTime)
    );
    const isRoot = attributes["genkit:isRoot"];
    if (!isRoot) {
      throw new GenkitError({
        status: "FAILED_PRECONDITION",
        message: "FeatureTelemetry tick called with non-root span."
      });
    }
    const state = attributes["genkit:state"];
    if (state === "success") {
      this.writeFeatureSuccess(name, latencyMs);
    } else if (state === "error") {
      const errorName = extractErrorName(span.events) || "<unknown>";
      this.writeFeatureFailure(name, latencyMs, errorName);
    } else {
      logger.warn(`Unknown state; ${state}`);
      return;
    }
    if (logIO) {
      const input = attributes["genkit:input"];
      const output = attributes["genkit:output"];
      if (input) {
        this.recordIO(span, "Input", name, path, input, projectId);
      }
      if (output) {
        this.recordIO(span, "Output", name, path, output, projectId);
      }
    }
  }
  writeFeatureSuccess(featureName, latencyMs) {
    const dimensions = {
      name: featureName,
      status: "success",
      source: "ts",
      sourceVersion: GENKIT_VERSION
    };
    this.featureCounter.add(1, dimensions);
    this.featureLatencies.record(latencyMs, dimensions);
  }
  writeFeatureFailure(featureName, latencyMs, errorName) {
    const dimensions = {
      name: featureName,
      status: "failure",
      source: "ts",
      sourceVersion: GENKIT_VERSION,
      error: errorName
    };
    this.featureCounter.add(1, dimensions);
    this.featureLatencies.record(latencyMs, dimensions);
  }
  recordIO(span, tag, featureName, qualifiedPath, input, projectId) {
    const path = toDisplayPath(qualifiedPath);
    const sharedMetadata = __spreadProps(__spreadValues({}, createCommonLogAttributes(span, projectId)), {
      path,
      qualifiedPath,
      featureName
    });
    logger.logStructured(`${tag}[${path}, ${featureName}]`, __spreadProps(__spreadValues({}, sharedMetadata), {
      content: input
    }));
  }
}
const featuresTelemetry = new FeaturesTelemetry();
export {
  featuresTelemetry
};
