// @genkit-ai/google-cloud/lib/gcpLogger.mjs
import {
  __async,
  __spreadProps,
  __spreadValues
} from "./chunk-E3LOUS7X.mjs";
import { LoggingWinston } from "@google-cloud/logging-winston";
import { getCurrentEnv } from "genkit";
import { logger } from "genkit/logging";
import { loggingDenied, loggingDeniedHelpText } from "./utils.js";
let additionalStream;
class GcpLogger {
  constructor(config) {
    this.config = config;
  }
  getLogger(env) {
    return __async(this, null, function* () {
      const winston = yield import("winston");
      const format = this.shouldExport(env) ? { format: winston.format.json() } : {
        format: winston.format.printf((info) => {
          return `[${info.level}] ${info.message}`;
        })
      };
      let transports = [];
      transports.push(
        this.shouldExport(env) ? new LoggingWinston({
          projectId: this.config.projectId,
          labels: { module: "genkit" },
          prefix: "genkit",
          logName: "genkit_log",
          credentials: this.config.credentials,
          autoRetry: true,
          defaultCallback: yield this.getErrorHandler()
        }) : new winston.transports.Console()
      );
      if (additionalStream) {
        transports.push(
          new winston.transports.Stream({ stream: additionalStream })
        );
      }
      return winston.createLogger(__spreadProps(__spreadValues({
        transports
      }, format), {
        exceptionHandlers: [new winston.transports.Console()]
      }));
    });
  }
  getErrorHandler() {
    return __async(this, null, function* () {
      let instructionsLogged = false;
      let helpInstructions = yield loggingDeniedHelpText();
      return (err) => __async(this, null, function* () {
        const defaultLogger = logger.defaultLogger;
        if (err && loggingDenied(err)) {
          if (!instructionsLogged) {
            instructionsLogged = true;
            defaultLogger.error(
              `Unable to send logs to Google Cloud: ${err.message}

${helpInstructions}
`
            );
          }
        } else if (err) {
          defaultLogger.error(`Unable to send logs to Google Cloud: ${err}`);
        }
        if (err) {
          logger.init(
            yield new GcpLogger(this.config).getLogger(getCurrentEnv())
          );
          defaultLogger.info("Initialized a new GcpLogger.");
        }
      });
    });
  }
  shouldExport(env) {
    return this.config.export;
  }
}
function __addTransportStreamForTesting(stream) {
  additionalStream = stream;
}
export {
  GcpLogger,
  __addTransportStreamForTesting
};
